### PR TITLE
refactor(ivy): remove styling state storage and introduce direct style writing

### PR DESCRIPTION
**Pull request by matsko**
This patch is a final major refactor in styling Angular.

This PR includes three main fixes:
- All temporary state taht is persisted between template style/class application
  and style/class application in host bindings is now removed.
- Removes the `styling()` and `stylingApply()` instructions.
- Introduces a "direct apply" mode that is used apply prop-based
  style/class in the event that there are no map-based bindings as
  well as property collisions.

### CODE REVIEW

**Comment by matsko** [comment]
Current Benchmarks:

```
[VE]
CREATE = time: 41.33235099332234 ms / iteration
CREATE_DIFF = 0%
UPDATE = time: 6.8551655627948245 ms / iteration
UPDATE_DIFF = 0%

[IVY - Master]
CREATE = time: 53.1655298012547 ms / iteration
CREATE_DIFF = 28%
UPDATE = time: 7.157582781371654 ms / iteration
UPDATE_DIFF = 4%

[IVY - THIS PR]
CREATE = time: 45.67341059578362 ms / iteration
CREATE_DIFF = 10%
UPDATE = time: 7.0579470197815874 ms / iteration
UPDATE_DIFF = 2%
```

**Comment by matsko** [comment]
Presubmit = http://test/OCL:266449721:BASE:266449752:1567198550650:eee361a5

Ivy Presubmit = http://test/OCL:266449721:BASE:266449878:1567198589588:d30e3cd1

**Comment by matsko** [comment]
# Benchmarks

## Large Table

### VE
CREATE = 38.435331126270256 ms
UPDATE = 6.078046357986943 ms

### IVY (based on master)
CREATE = 54.54857615973347 ms (40% slower)
UPDATE = 7.8921523179345465 ms (16% slower)

### IVY (based on this PR)
CREATE = 46.36072847700758 ms (20% slower)
UPDATE = 7.016092714709415 ms (16% slower)

## Micro Benchmarks

### [style.prop] (result = 45% faster)
MASTER = 20.81s user 0.21s system 103% cpu 20.339 total
THIS PR = 11.33s user 0.08s system 100% cpu 11.372 total

### [style.prop] and [class.prop] (result = 28% faster)
MASTER = 30.63s user 0.27s system 102% cpu 30.244 total
THIS PR = 21.80s user 0.11s system 100% cpu 21.882 total

### [style] and [class] (result = same)
MASTER = 68.91s user 1.22s system 104% cpu 1:07.15 total
THIS PR = 67.72s user 1.07s system 102% cpu 1:06.83 total

File: packages/core/src/render3/instructions/select.ts [hunk 1]
```diff
@@ -41,6 +42,8 @@ export function selectInternal(lView: LView, index: number, checkNoChangesMode:
   ngDevMode && assertGreaterThan(index, -1, 'Invalid index');
   ngDevMode && assertDataInRange(lView, index + HEADER_OFFSET);
 
+  executeElementExitFn();
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions/select.ts]: This is a rather expensive operation in the perf-critical path. Can you think of ways of guarding this logic so we don't need to do this call in case there is no styling?

File: packages/core/src/render3/instructions/select.ts [hunk 2]
```diff
@@ -64,4 +67,7 @@ export function selectInternal(lView: LView, index: number, checkNoChangesMode:
   // state. If we run `setSelectedIndex` *before* we run the hooks, in some cases the selected index
   // will be altered by the time we leave the `ɵɵselect` instruction.
   setSelectedIndex(index);
+
+  resetStylingState();
+  resetActiveElementFlags();
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions/select.ts]: This is a rather expensive operation in the perf-critical path. Can you think of ways of guarding this logic so we don't need to do this call in case there is no styling?

File: packages/core/src/render3/state.ts [hunk 3]
```diff
@@ -164,13 +186,45 @@ let activeDirectiveSuperClassHeight = 0;
...
+ * instruction is called or when the template or host bindings instruction
+ * code is invoked. The element is then "exited" when the next element is
+ * selected or when change detection for the template or host bindings is
+ * complete. When this occurs (the element change operation) then an exit
+ * function will be invoked if it has been set. This function can be used
+ * to assign that exit function.
+ *
+ * @param fn
+ */
+export function setElementExitFn(fn: Function): void {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/state.ts]: So, the way I understand it is that we are setting a global state here to have a function to execute and the element "end".  Also, my understanding is that a function to execute could be figured out on the first template pass and stored on TNode.

I would suggest that we:
* introduce a new field on TView / TNode holding those functions to execute;
* guard execution of those functions with a TNode flag
* remove global state.

Obviously I'm missing toons of details here and what I'm suggesting might not be doable or trivial - at this point this is a broad-strokes suggestion, to be refined.

File: packages/core/src/render3/instructions/select.ts [hunk 4]
```diff
@@ -64,4 +69,10 @@ export function selectInternal(lView: LView, index: number, checkNoChangesMode:
   // state. If we run `setSelectedIndex` *before* we run the hooks, in some cases the selected index
   // will be altered by the time we leave the `ɵɵselect` instruction.
   setSelectedIndex(index);
+
+  if (hasActiveElementFlag(ActiveElementFlags.ResetStylesOnExit)) {
```
Comment (mhevery) [inline packages/core/src/render3/instructions/select.ts]: per our discussion once flags are folded into `_selectedIndex` this reseting becomes unnecessary.

File: packages/core/src/render3/instructions/shared.ts [hunk 5]
```diff
@@ -500,7 +503,9 @@ function executeTemplate<T>(
     }
     templateFn(rf, context);
   } finally {
+    executeElementExitFn();
```
Comment (mhevery) [inline packages/core/src/render3/instructions/shared.ts]:
```suggestion
    if (hasActiveElementFlag(ActiveElementFlags.RunExitFn)) {
      executeElementExitFn();
    }
```

File: packages/core/src/render3/state.ts [hunk 6]
```diff
@@ -129,31 +129,56 @@ export function getLView(): LView {
...
+}
+
+let _selectedIndexFlags = ActiveElementFlags.Initial;
+
+/**
+ * Clears all active element flags.
+ */
+export function resetActiveElementFlags() {
+  if (_selectedIndexFlags !== ActiveElementFlags.Initial) {
+    _selectedIndexFlags = ActiveElementFlags.Initial;
```
Comment (mhevery) [inline packages/core/src/render3/state.ts]: per our discussion store the state in `_selectedIndex`

File: packages/core/src/render3/styling_next/util.ts [hunk 7]
```diff
@@ -197,17 +218,22 @@ export function getStylingMapArray(value: TStylingContext | StylingMapArray | nu
...
-  return Array.isArray(value) &&
-      value.length >= TStylingContextIndex.MapBindingsBindingsStartPosition &&
+  return Array.isArray(value) && value.length >= TStylingContextIndex.ValuesStartPosition &&
       typeof value[1] !== 'string';
 }
 
+export function isStylingMapArray(value: TStylingContext | StylingMapArray | null): boolean {
+  // the StylingMapArray is in the format of [initial, prop, string, prop, string]
+  // and this is the defining value to distinguish between arrays
+  return Array.isArray(value) && isStylingContext(value) === false;
```
Comment (mhevery) [inline packages/core/src/render3/styling_next/util.ts]: this code will check `isArray` twice which is not necessary.

File: packages/core/src/render3/styling_next/util.ts [hunk 8]
```diff
@@ -225,6 +251,12 @@ export function getMapProp(map: StylingMapArray, index: number): string {
   return map[index + StylingMapArrayIndex.PropOffset] as string;
 }
 
+const MAP_DIRTY_VALUE = {};
```
Comment (mhevery) [inline packages/core/src/render3/styling_next/util.ts]:
```suggestion
const MAP_DIRTY_VALUE = typeof ngDevMode !== 'undefined' && ngDevMode ? {}: {MAP_DIRTY_VALUE:true };
```
This makes debugging easier.

File: packages/core/src/render3/styling_next/interfaces.ts [hunk 9]
```diff
@@ -283,94 +323,94 @@ import {LView} from '../interfaces/view';
...
-   * template-level as well as within host bindings on the same element. This is a
-   * rare case, and a storage map is required so that the state values can be restored
-   * between the template code running and the host binding code executing.
+   * Whether or not there are map-based and prop-based bindings present
    */
-  PersistStateValues = 0b10,
+  HasPropAndMapBindings = 0b0000011,
+
+  /**
+   * Whether or not there are two or more sources for a single property in the context
```
Comment (mhevery) [inline packages/core/src/render3/styling_next/interfaces.ts]: It would be nice if you would include a code example demonstrating when this flag would be triggered. Otherwise "two or more source for a single property" does not mean much to a reader

(Do same for the rest of the flags)

File: packages/core/src/render3/styling_next/bindings.ts [hunk 10]
```diff
@@ -316,129 +326,99 @@ function allocateNewContextEntry(
...
+    applyStylingViaContext(
+        classesContext, renderer, element, data, state.classesBitMask, setClass, null,
+        hostBindingsMode);
   }
-  return allowFlush;
-}
 
-function contextHasUpdates(context: TStylingContext | null, bitMask: number) {
-  return context && bitMask > BIT_MASK_START_VALUE;
+  resetStylingState();
```
Comment (mhevery) [inline packages/core/src/render3/styling_next/bindings.ts]: Is this necessary?  I would assume that the next `select` instruction would take care of the reset.

File: integration/_payload-limits.json [hunk 11]
```diff
@@ -26,15 +26,6 @@
       }
     }
   },
-  "cli-hello-world-ivy-i18n": {
```
Comment (mhevery) [inline integration/_payload-limits.json]: Why is this removed?

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/core/test/bundling/hello_world/bundle.golden_symbols.json [hunk 12]
```diff
@@ -479,6 +491,9 @@
   {
     "name": "syncViewWithBlueprint"
   },
+  {
+    "name": "unsetActiveElementFlag"
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/core/test/bundling/hello_world/bundle.golden_symbols.json]: Done

File: integration/_payload-limits.json [hunk 13]
```diff
@@ -21,20 +21,11 @@
     "master": {
       "uncompressed": {
         "runtime": 1440,
-        "main": 123904,
+        "main": 125448,
```
[the remark this answers is not in the document]
Comment (matsko) [reply integration/_payload-limits.json]: The payload limit amounts were off (it was rebased many times) and the stylingApply code isn't gone (we still flush things like before), but the instruction is now called implicitly after one or more styling instructions is executed.

[the remark this answers is not in the document]
Comment (kara) [reply integration/_payload-limits.json]: By gone, I meant from the generated code :-)  But yeah, that makes sense.

File: packages/core/src/render3/state.ts [hunk 14]
```diff
@@ -185,26 +170,21 @@ export function unsetActiveElementFlag(flag: ActiveElementFlags) {
...
       resetStylingState();
     }
-    resetActiveElementFlags();
+    setSelectedIndex(elementIndex === null ? -1 : elementIndex);
+    activeDirectiveId = 0;
   }
 }
 
 let _elementExitFn: Function|null = null;
 export function executeElementExitFn() {
```
Comment (mhevery) [inline packages/core/src/render3/state.ts]: Can we get rid of `executeElementExitFn` since all it does is just call `_elementExitFn`?

File: packages/core/src/render3/instructions/shared.ts [hunk 15]
```diff
@@ -500,6 +503,12 @@ function executeTemplate<T>(
     }
     templateFn(rf, context);
   } finally {
+    if (hasActiveElementFlag(ActiveElementFlags.RunExitFn)) {
+      executeElementExitFn();
+    }
+    if (hasActiveElementFlag(ActiveElementFlags.ResetStylesOnExit)) {
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/core/src/render3/instructions/shared.ts]: If there is an exit function then the styles need to also be reset. The case for this is that there is a final element that runs before a templateFn finishes (no final `select()` there). 

Note that this will be removed in the next PR (there will only be an exitFn that will handle both the reset and apply).

File: packages/core/src/sanitization/bypass.ts [hunk 16]
```diff
@@ -89,10 +87,10 @@ class SafeResourceUrlImpl extends SafeValueImpl implements SafeResourceUrl {
   getTypeName() { return BypassType.ResourceUrl; }
 }
 
-export function unwrapSafeValue(value: SafeValue): string {
+export function unwrapSafeValue(value: any, backupValue: string = ''): string {
```
Comment (mhevery) [inline packages/core/src/sanitization/bypass.ts]: `backupValue` is always `value`. Can we just drop the second argument?

**Review by mhevery — APPROVED** [review]
with minor comments

File: packages/core/src/render3/styling_next/util.ts [hunk 17]
```diff
@@ -21,124 +44,122 @@ const TEMPLATE_DIRECTIVE_INDEX = 0;
...
+export function hasConfig(context: TStylingContext, flag: TStylingConfig, bitsMustMatch?: boolean) {
+  const value = getConfig(context) & flag;
+  return bitsMustMatch ? value === flag : value !== 0;
+}
+
+/**
+ * Determines whether or not to apply styles/classes directly or via context resolution.
+ */
+export function allowDirectStyling(context: TStylingContext, hostBindingsMode: boolean): boolean {
+  return isContextLocked(context, hostBindingsMode) &&
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/core/src/render3/styling_next/util.ts]: Yes. We should check the config value once.

File: packages/core/src/render3/instructions/select.ts (hunk 4 above)
Comment (matsko) [reply packages/core/src/render3/instructions/select.ts]: Fixed.

File: packages/core/src/render3/instructions/shared.ts (hunk 5 above)
Comment (matsko) [reply packages/core/src/render3/instructions/shared.ts]: Fixed.

File: packages/core/src/render3/state.ts (hunk 6 above)
Comment (matsko) [reply packages/core/src/render3/state.ts]: Fixed.

File: packages/core/test/bundling/hello_world/bundle.golden_symbols.json [hunk 18]
```diff
@@ -420,7 +432,7 @@
     "name": "renderView"
   },
   {
-    "name": "resetAllStylingState"
+    "name": "resetActiveElementFlags"
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/core/test/bundling/hello_world/bundle.golden_symbols.json]: Removed

File: packages/core/src/render3/styling_next/util.ts [hunk 19]
```diff
@@ -253,3 +284,134 @@ export function forceStylesAsString(styles: {[key: string]: any} | null | undefi
   }
   return str;
 }
+
+export function normalizeStylingValue(value: any) {
+  return value instanceof SafeValueImpl ? unwrapSafeValue(value) : value;
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/core/src/render3/styling_next/util.ts]: Updated

File: packages/core/src/render3/styling_next/util.ts [hunk 20]
```diff
@@ -21,124 +44,121 @@ const TEMPLATE_DIRECTIVE_INDEX = 0;
...
 
-function getConfig(context: TStylingContext) {
+export function getConfig(context: TStylingContext) {
   return context[TStylingContextIndex.ConfigPosition];
 }
 
-export function setConfig(context: TStylingContext, value: number) {
+export function hasConfig(context: TStylingContext, flag: TStylingConfig, bitsMustMatch?: boolean) {
+  const value = getConfig(context) & flag;
+  return bitsMustMatch ? value === flag : value > 0;
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/core/src/render3/styling_next/util.ts]: Fixed

File: packages/core/src/render3/styling_next/util.ts (hunk 7 above)
Comment (matsko) [reply packages/core/src/render3/styling_next/util.ts]: Fixed

File: packages/core/src/render3/styling_next/util.ts (hunk 8 above)
Comment (matsko) [reply packages/core/src/render3/styling_next/util.ts]: Updated

File: packages/core/src/render3/styling_next/interfaces.ts (hunk 9 above)
Comment (matsko) [reply packages/core/src/render3/styling_next/interfaces.ts]: Updated

File: packages/core/src/render3/styling_next/bindings.ts (hunk 10 above)
Comment (matsko) [reply packages/core/src/render3/styling_next/bindings.ts]: Yes this is necessary. The `stylingState` has nothing to do with the `select` bits.

File: packages/core/src/render3/state.ts (hunk 14 above)
Comment (matsko) [reply packages/core/src/render3/state.ts]: No we can't because you can't export a single variable while having that variable be changed elsewhere.

File: packages/core/src/sanitization/bypass.ts (hunk 16 above)
Comment (matsko) [reply packages/core/src/sanitization/bypass.ts]: I do not understand what you mean here.

File: aio/scripts/_payload-limits.json [hunk 21]
```diff
@@ -28,7 +28,7 @@
       "uncompressed": {
         "runtime-es5": 2932,
         "runtime-es2015": 2938,
-        "main-es5": 550854,
+        "main-es5": 552068,
```
Comment (kara) [inline aio/scripts/_payload-limits.json]: Do we know what's causing the size increase?

File: packages/core/test/render3/perf/style_binding/index.ts [hunk 22]
```diff
@@ -30,69 +30,39 @@ import {setupRootViewWithEmbeddedViews} from '../setup';
...
-    ɵɵadvance(1);
+    ɵɵadvance(8);
     ɵɵstyleProp('background-color', 'color8');
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(9);
     ɵɵstyleProp('background-color', 'color9');
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(10);
```
Comment (kara) [inline packages/core/test/render3/perf/style_binding/index.ts]: The index changes here don't look right. `advance` instructions should be relative, so this would be taking you to index ~55. Can you revert the index changes?

File: packages/core/test/render3/perf/style_and_class_bindings/index.ts [hunk 23]
```diff
@@ -31,89 +31,50 @@ import {setupRootViewWithEmbeddedViews} from '../setup';
...
     ɵɵstyleProp('width', '700px');
     ɵɵclassProp('scale', true);
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(9);
     ɵɵstyleProp('width', '800px');
     ɵɵclassProp('scale', true);
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(10);
```
Comment (kara) [inline packages/core/test/render3/perf/style_and_class_bindings/index.ts]: Same here - the index changes are invalid

File: packages/core/test/render3/perf/map_based_style_and_class_bindings/index.ts [hunk 24]
```diff
@@ -30,79 +30,49 @@ import {setupRootViewWithEmbeddedViews} from '../setup';
...
     ɵɵstyleMap({width: '70px', height: '700px'});
     ɵɵclassMap('one two');
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(9);
     ɵɵstyleMap({width: '80px', height: '800px'});
     ɵɵclassMap('one two');
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(10);
```
Comment (kara) [inline packages/core/test/render3/perf/map_based_style_and_class_bindings/index.ts]: Fix here too

File: packages/core/src/sanitization/bypass.ts [hunk 25]
```diff
@@ -89,10 +87,10 @@ class SafeResourceUrlImpl extends SafeValueImpl implements SafeResourceUrl {
   getTypeName() { return BypassType.ResourceUrl; }
 }
 
-export function unwrapSafeValue(value: SafeValue): string {
+export function unwrapSafeValue(value: any): string {
```
Comment (kara) [inline packages/core/src/sanitization/bypass.ts]: Curious why this is necessary

File: packages/core/src/render3/state.ts [hunk 26]
```diff
@@ -129,31 +129,38 @@ export function getLView(): LView {
...
  * value is reserved for the template.
  */
-const MIN_DIRECTIVE_ID = 1;
+let activeDirectiveId = 0;
 
-let activeDirectiveId = MIN_DIRECTIVE_ID;
+/**
+ * Flags used for an active element during change detection.
+ *
+ * These flags are used to signal whether to run any sub routine code when
```
Comment (kara) [inline packages/core/src/render3/state.ts]: What does subroutine code mean here? Example would be helpful

```suggestion
 * These flags are used to signal whether to run any sub routine code when
```

**Review by kara — CHANGES_REQUESTED** [review]

File: packages/core/test/render3/perf/style_binding/index.ts (hunk 22 above)
Comment (matsko) [reply packages/core/test/render3/perf/style_binding/index.ts]: Fixed already

File: packages/core/test/render3/perf/style_and_class_bindings/index.ts (hunk 23 above)
Comment (matsko) [reply packages/core/test/render3/perf/style_and_class_bindings/index.ts]: Fixed

File: packages/core/test/render3/perf/map_based_style_and_class_bindings/index.ts (hunk 24 above)
Comment (matsko) [reply packages/core/test/render3/perf/map_based_style_and_class_bindings/index.ts]: Fixed

File: packages/core/src/sanitization/bypass.ts (hunk 25 above)
Comment (matsko) [reply packages/core/src/sanitization/bypass.ts]: Me and Misko discussed this. The end result is that the return of the empty string is the wrong behavior. It should just unwrap the value if possible

File: packages/core/src/render3/state.ts (hunk 26 above)
Comment (matsko) [reply packages/core/src/render3/state.ts]:
```
 * These flags are used within other instructions to inform cleanup or
 * exit operations to run when an element is being processed
```

Comment (kara) [reply packages/core/src/render3/state.ts]: 👍

File: aio/scripts/_payload-limits.json (hunk 21 above)
Comment (kara) [reply aio/scripts/_payload-limits.json]: @matsko and I discussed. He will look into this and fix the size regression as a follow-up PR.

**Review by kara — APPROVED** [review]
LGTM
